### PR TITLE
chore: bump to Java 11

### LINKED ISSUE #1186: Bump Java version to Java 11
Related #371.

Need to be careful with `Nopol`.

### PR DESCRIPTION
**Pull request by andre15silva**
Closes #1186

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -8,7 +8,7 @@ kind: Pod
 spec:
   containers:
   - name: maven
-    image: repairnator/ci-env:latest
+    image: andre15silva/repairnator-ci:latest
```
Comment (monperrus) [inline Jenkinsfile:11]: added you to https://hub.docker.com/orgs/repairnator such that you can push a new version to the organization instead.

File: src/repairnator-pipeline/src/test/java/fr/inria/spirals/repairnator/process/step/repair/TestNopolRepair.java [hunk 2]
```diff
@@ -58,7 +59,8 @@ public void tearDown() throws IOException {
         FileHelper.deleteFile(tmpDir);
     }
 
-    @Test
+    @Ignore
+    //@Test
```
Comment (monperrus) [inline src/repairnator-pipeline/src/test/java/fr/inria/spirals/repairnator/process/step/repair/TestNopolRepair.java:63]: avoid comments + @Ignore is enough. could you add a comment to explain why it is ignored?

**Review by monperrus — COMMENTED** [review]
Very nice, very important PR, thanks for it.

- see comments
- Why is CI failing?

File: src/repairnator-pipeline/src/test/java/fr/inria/spirals/repairnator/process/step/repair/TestNopolRepair.java (hunk 2 above)
Comment (andre15silva) [reply src/repairnator-pipeline/src/test/java/fr/inria/spirals/repairnator/process/step/repair/TestNopolRepair.java:63]: Only `@Ignore` didn't work for some reason. I commented `@Test` just to make it skip. I'll debug that.

File: Jenkinsfile (hunk 1 above)
Comment (andre15silva) [reply Jenkinsfile:11]: Ok, thanks! Will do that.

**Comment by andre15silva** [comment]
>     * Why is CI failing?

I'm debugging that. Just opened this PR to have the CI env running to check in parallel.

**Comment by monperrus** [comment]
Now that we fixed the Sorald issue, what about finishing the move to Java 11?

**Comment by andre15silva** [comment]
> Now that we fixed the Sorald issue, what about finishing the move to Java 11?

Yup, I'm gonna work on that this afternoon. Hopefully it will be finished today.

**Comment by andre15silva** [comment]
We need to deploy a new version of `maven-repair`, that supports Java 11, before upgrading the `pipeline`, as `NPERepair` depends on it still (changes with #1172).
This is the major cause of failures, not being able to access the new version.

Also, what does `TestGlobalPatchAnalysis.testODSPatchClassification` do exactly? Is it only related to Nopol? It is failing because Nopol requires Java 8.

Same thing goes for `TestProjectInspector.testPatchFailingProject`, but I assume we can just remove `Nopol` from the repair tools for that test.

**Comment by andre15silva** [comment]
@monperrus do you think we can deploy `maven-repair` as it is in this PR, so that we can use that version as a dependency in the `pipeline`?

**Comment by javierron** [comment]
Hey @andre15silva

> Also, what does TestGlobalPatchAnalysis.testODSPatchClassification do exactly? Is it only related to Nopol? It is failing because Nopol requires Java 8.

I think that the goal of that test is to check that ODS classification is applied to the generated patches, regardless of the tool used to generate the patches, so replacing Nopol for other tool is not a problem.

> Same thing goes for TestProjectInspector.testPatchFailingProject, but I assume we can just remove Nopol from the repair tools for that test.

The same applies here, the goal of the test seems to be to check that all steps, serializers, etc. are correct. However to remove Nopol from both tests we would have to provide a tool/project combination that creates a patch (maybe introducing a NPE in the test project?).

Also, I have a question: Is the goal to move the project to a java 11 only version? If that is true why not remove Nopol completely? Or do we want to keep the project compatible with both java 8 and 11 simultaneously?

Please tell me if I can help on something to get this merged, I would like to rebase my work to this, to avoid any possible java version issues.

**Comment by monperrus** [comment]
> @monperrus do you think we can deploy maven-repair

Do you mean before this PR? If "yes", I'd be happy to redeploy the latest version.

**Comment by andre15silva** [comment]
> Do you mean before this PR? If "yes", I'd be happy to redeploy the latest version.

Yes. If you can deploy the version before this PR with the profile `noNopol` activated that would be great. Or as it sits in this PR, which is with that profile active by default (and Nopol tests deactivated).

**Comment by javierron** [comment]
> Do you know where I can find documentation/information on ODS classification? 

This is part of the Coming repository, you can check the docs [here](https://github.com/SpoonLabs/coming/blob/master/docs/ODS.md)

**Comment by andre15silva** [comment]
> This is part of the Coming repository, you can check the docs [here](https://github.com/SpoonLabs/coming/blob/master/docs/ODS.md?rgh-link-date=2021-02-22T19%3A16%3A33Z)

Thanks!

**Comment by monperrus** [comment]
> If I'm not mistaken, the goal is to drop support for Java 8

yes!

What remains to be done here in order to merge?

**Comment by andre15silva** [comment]
> What remains to be done here in order to merge?

The deployment of `maven-repair` with `noNopol` profile activated. Without that the pipeline can't use NPEFix.

**Comment by monperrus** [comment]
Just tried to deploy maven-repair.

It fails, because maven-repair does not compile

```
$ mvn clean deploy -PnoNopol
[ERROR] COMPILATION ERROR : 
[INFO] -------------------------------------------------------------
[ERROR] /home/martin/martin-no-backup/repairnator/src/maven-repair/src/main/java/com/github/tdurieux/repair/maven/plugin/NopolMojo.java:[30,1] cannot find symbol
  symbol:   static checkToolsJar
  location: class
[ERROR] /home/martin/martin-no-backup/repairnator/src/maven-repair/src/main/java/com/github/tdurieux/repair/maven/plugin/NopolMojo.java:[62,13] cannot find symbol
  symbol:   method checkToolsJar()
  location: class com.github.tdurieux.repair.maven.plugin.NopolMojo
```

Are you able to compile the latest master of maven-repair with -PnoNopol? How can a compilation failure happen given that we have CI?

Thanks!

**Comment by andre15silva** [comment]
> Just tried to deploy maven-repair.
> 
> It fails, because maven-repair does not compile
> 
> ```
> $ mvn clean deploy -PnoNopol
> [ERROR] COMPILATION ERROR : 
> [INFO] -------------------------------------------------------------
> [ERROR] /home/martin/martin-no-backup/repairnator/src/maven-repair/src/main/java/com/github/tdurieux/repair/maven/plugin/NopolMojo.java:[30,1] cannot find symbol
>   symbol:   static checkToolsJar
>   location: class
> [ERROR] /home/martin/martin-no-backup/repairnator/src/maven-repair/src/main/java/com/github/tdurieux/repair/maven/plugin/NopolMojo.java:[62,13] cannot find symbol
>   symbol:   method checkToolsJar()
>   location: class com.github.tdurieux.repair.maven.plugin.NopolMojo
> ```
> 
> Are you able to compile the latest master of maven-repair with -PnoNopol? How can a compilation failure happen given that we have CI?

`mvn clean compile -PnoNopol` works for me. 

Did you install `repairnator-core` before? That static method `checkToolsJar` is in `repairnator-core`.

**Comment by monperrus** [comment]
You're right. The latest snapshot of core was from November (because CD is broken, see #1167 )

Delivered a new core 3-3-SNAPSHOT manually.

Just deployed version `0bf3bd6ffadf3638ee604b83053426c739aff1f7` to Maven Central. Note that it depends on a snapshot (3-3-SNAPSHOT) which is a bad practice (releases should never depend on snapshots).

But for now, we can proceed with `0bf3bd6ffadf3638ee604b83053426c739aff1f7`, which will soon appear at https://repo.maven.apache.org/maven2/fr/inria/gforge/spirals/repair-maven-plugin/

**Comment by andre15silva** [comment]
Thanks! I'll get this working with that version. 

Note that this is just temporary, when #1172 is merged it will no longer be necessary.

**Comment by monperrus** [comment]
great, this move to Java 11 will be a serious milestone!

**Comment by andre15silva** [comment]
Is it normal to take this long to appear in the repository?

**Comment by monperrus** [comment]
release at last successful: https://repo1.maven.org/maven2/fr/inria/gforge/spirals/repair-maven-plugin/0bf3bd6ffadf3638ee604b83053426c739aff1f7/

**Comment by andre15silva** [comment]
There is now a problem with `gzoltar`.

When using `mvn fr.inria.gforge.spirals:repair-maven-plugin:0bf3bd6ffadf3638ee604b83053426c739aff1f7:npefix` on a project, we get this error:

```
[DEBUG] Using transporter WagonTransporter with priority -1.0 for http://sachaproject.gforge.inria.fr/repositories/releases/
[DEBUG] Using connector BasicRepositoryConnector with priority 0.0 for http://sachaproject.gforge.inria.fr/repositories/releases/
Downloading from gforge.inria.fr-release: http://sachaproject.gforge.inria.fr/repositories/releases/com/gzoltar/gzoltar/0.1.1/gzoltar-0.1.1.pom
Progress (1): gzoltar-0.1.1.pom (462 B)
[WARNING] Could not validate integrity of download from http://sachaproject.gforge.inria.fr/repositories/releases/com/gzoltar/gzoltar/0.1.1/gzoltar-0.1.1.pom
org.eclipse.aether.transfer.ChecksumFailureException: Checksum validation failed, no checksums available
```

This happens because there is no checksum available for this version of gzoltar on this repository, and maven central does not have this version available to serve as backup. 
This doesn't show up when installing `maven-repair` tho, it doesn't check it for some reason I haven't figured out.

I see three ways of fixing this:

- Moving `gzoltar` dependency to a more recent version that is available in maven central, and then re-deploying `maven-repair`.
- Updating the `gzoltar` deploy so it has the checksums.
- Merging #1172 before, so that the pipeline does all the work and doesn't need to call `maven-repair`.

I think the later two are probably the better options.


@monperrus WDYT?

**Comment by monperrus** [comment]
Argh. Yes, GZoltar is one the known legacy problems we have, see https://github.com/eclipse/repairnator/issues/1029

I would prefer to merge #1172 first then. It looks good overall and CI is green. What remains to be done there?

**Comment by andre15silva** [comment]
> I would prefer to merge #1172 first then. It looks good overall and CI is green. What remains to be done there?

If we set aside the scope issue as discussed, it is ready to be merged.

**Comment by andre15silva** [comment]
## Changelog:
- Move CI docker image to Java 11
- Activate `noNopol` profile by default on `maven-repair` and `repairnator-pipeline` so that we don't load `tools.jar`
- Change synthesis and localizer default params to SMT and CoCoSpoon respectively, so as to support Java 11.
- Add z3 binaries to test resources in `maven-repair` to enable SMT
- Update `maven-jacoco-plugin` to a Java 11 compatible version
- Change `gitrepourl` of `AstorJKali` tests to the same project but compatible with Java 11
- Update test assertions due to changes in Java version and Nopol parameters

**Comment by andre15silva** [comment]
Ready for review! @monperrus

File: src/repairnator-pipeline/src/test/java/fr/inria/spirals/repairnator/pipeline/TestPipeline.java [hunk 3]
```diff
@@ -72,7 +72,7 @@ public void testPipelineOnlyGitRepository() throws Exception {
     public void testPipelineGitRepositoryAndBranch() throws Exception {
         GitRepositoryLauncher l = new GitRepositoryLauncher(new String[]{
                 "--gitrepo",
-                "--gitrepourl", "https://github.com/surli/failingProject",
```
Comment (monperrus) [inline src/repairnator-pipeline/src/test/java/fr/inria/spirals/repairnator/pipeline/TestPipeline.java:75]: What about using https://github.com/repairnator/failingProject/ instead? (sent you an invitation)

File: src/repairnator-pipeline/src/test/java/fr/inria/spirals/repairnator/pipelineb/TestPipelinebGithubMode.java [hunk 4]
```diff
@@ -52,7 +52,7 @@ public void testPipelineOnlyGitRepository() throws Exception {
     public void testPipelineGitRepositoryAndBranch() throws Exception {
         GithubMainProcess mainProc = (GithubMainProcess) MainProcessFactory.getGithubMainProcess(new String[]{
                 "--gitrepo",
-                "--gitrepourl", "https://github.com/surli/failingProject",
+                "--gitrepourl", "https://github.com/andre15silva/failingProject",
```
Comment (monperrus) [inline src/repairnator-pipeline/src/test/java/fr/inria/spirals/repairnator/pipelineb/TestPipelinebGithubMode.java:55]: idem

File: src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/process/step/repair/nopol/AbstractNopolRepair.java [hunk 5]
```diff
@@ -134,7 +135,7 @@ protected void runNopol(Set<FailureLocation> failureLocation, List<String> tests
         nopolContext.setMaxTimeInMinutes(timeout);
         nopolContext.setLocalizer(NopolContext.NopolLocalizer.COCOSPOON);
         nopolContext.setSolverPath(this.getConfig().getZ3solverPath());
-        nopolContext.setSynthesis(NopolContext.NopolSynthesis.DYNAMOTH);
```
Comment (monperrus) [inline src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/process/step/repair/nopol/AbstractNopolRepair.java:137]: could you add a comment to explain why we cannot use DYNAMOTH?

**Review by monperrus — COMMENTED** [review]
Waoow, it looks really good @andre15silva ! See my last comments about the testing infrastructure and then we're ready to officially move to Java 11.

File: src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/process/step/repair/nopol/AbstractNopolRepair.java (hunk 5 above)
Comment (andre15silva) [reply src/repairnator-pipeline/src/main/java/fr/inria/spirals/repairnator/process/step/repair/nopol/AbstractNopolRepair.java:137]: Done!

File: src/repairnator-pipeline/src/test/java/fr/inria/spirals/repairnator/pipeline/TestPipeline.java (hunk 3 above)
Comment (andre15silva) [reply src/repairnator-pipeline/src/test/java/fr/inria/spirals/repairnator/pipeline/TestPipeline.java:75]: Done!

**Comment by andre15silva** [comment]
Ready after CI runs again

**Comment by monperrus** [comment]
Thanks a lot @andre15silva for this foundational contribution.

**Pull request merged**